### ISSUE TITLE
Copper test bputest3 fails

### ISSUE THREAD

**Issue by dirkwhoffmann** [dirkwhoffmann/vAmiga#73]
Copper list: 

    ; Switch off biplanes in Copper cycle 80. Switch them on again in the next line
	dc.w	$5081,$FFFE  ; WAIT 
	dc.w    BPLCON0, $200 ; All bitplanes off
	dc.w	$5101,$FFFE  ; WAIT 
	dc.w    BPLCON0, (SCREEN_BIT_DEPTH<<12)|$200 ; Bitplanes on

	dc.l	$fffffffe
	dc.l	$fffffffe

Reference (SAE): 

<img width="712" alt="Screenshot 2019-06-25 at 18 18 15" src="https://user-images.githubusercontent.com/12561945/60115301-e88c7e00-9775-11e9-90da-4a5be6088cde.png">

vAmiga 🙈:

<img width="929" alt="Screenshot 2019-06-25 at 18 15 05" src="https://user-images.githubusercontent.com/12561945/60115329-f6da9a00-9775-11e9-9d37-024531465272.png">

**Comment by dirkwhoffmann** [comment]
Hypothesis: DMA slots in the currently processed fetch unit are never deleted (based on the observation that the coloured squares remain being aligned in SAE).

**Comment by naTmeg** [comment]
Maybe the comments on the WinUAE copper-code can help, which I stripped from SAE:
https://github.com/tonioni/WinUAE/blob/master/custom.cpp (starting at line 6922)

**Comment by dirkwhoffmann** [comment]
> Maybe the comments on the WinUAE copper-code can help

Thanks, Rupert, that helps! My favorite comment line is right at the beginning:

> /* The copper code.  The biggest nightmare in the whole emulator.

😂

Interestingly, writing this emulator has become like real science now. I make an observation, build up a hypotheses, and run an experiment (in SAE) to support it. It's a slow process, but I'm coming closer 😎:

<img width="929" alt="Screenshot 2019-06-25 at 19 06 48" src="https://user-images.githubusercontent.com/12561945/60119170-170e5700-977e-11e9-8884-baeb67ba4252.png">

But I'm sure my next experiment will dissolve my hypothesis into thin air. Just like real science.

Just a tiny question about SAE: Is it possible to tell SAE to remember the Kick ROM it is supposed to load? I looked through config.js, but didn't find a place like where I could just insert a default name. For all other options, I simply edited config.js to get the default values I need (like OCS chipset and stuff).

**Comment by naTmeg** [comment]
>Is it possible to tell SAE to remember the Kick ROM it is supposed to load?

No, that should not be possible in JS generally, because that would be a huge security risk. Imagine a script on some site that can load any local it wants. So, the user always have to specify a file manually.

But, you can dump any kick-rom into a 512k-UInt8Array (let's call it "kickdata12") stored in a separate file and fill the following manually:
cfg.memory.rom.name = "kick12.rom"; /* not really needed */
cfg.memory.rom.data = kickdata12;
cfg.memory.rom.size = kickdata12.length;
cfg.memory.rom.crc32 = crc32(kickdata12); /* not really needed */

I know the copper- and bitplane-code is very time consuming. I was in the same situation like you when I wrote SAE 0.1, but decided to port from other sources later, simply because I ran out of time and literally, out of money :( I worked >16h/day on it but had to start some job these days, to pay my bills. Like you, I love finding out things by myself :)

**Comment by dirkwhoffmann** [comment]
> Maybe the comments on the WinUAE copper-code can help, which I stripped from SAE:

That actually led me to the solution 😀. After drawing some diagrams of the sort

    With 5 planes, hitting at $7E:

    ; Set BPU to a certain number 
	dc.w	$4501,$FFFE  ; WAIT 
	dc.w    BPLCON0, (5<<12)|$200 ; Bitplanes on
	dc.w    $1fe, $0  ; NOP (Dummy write to unused register)

	; Emulate a Copper MOVE in the middle of the rasterline
	dc.w	$457F,$FFFE  ; WAIT 
	dc.w    BPLCON0, (5<<12)|$200 ; Bitplanes on
	dc.w    $1fe, $1  ; NOP (Dummy write to unused register)

	dc.l	$fffffffe
    55555555555555556666666666666666777777777777777788888888888888889999999999999999
    0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF0123456789ABCDEF
    .L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL.L.L.LLL
    .4.2.351.4.2.351.4.2.351.4.2.351.4.2.351.4.2.351.4.2.351.4.2.351.4.2.351.4.2.351
                                                ^   ^ ^ ^   ^
                                                |   | | |   |
                                                |   | | |   READ2
                                                |   | | READ1
                                                |   | READ2, pokeBPLCON0
                                                |   READ1
                                                | 
                                                WAKEUP

I got it right (I think): 

<img width="736" alt="Screenshot 2019-06-25 at 21 49 04" src="https://user-images.githubusercontent.com/12561945/60281421-34b9f880-9905-11e9-8d23-4e28f7d7f47e.png">

<img width="929" alt="Screenshot 2019-06-27 at 17 05 24" src="https://user-images.githubusercontent.com/12561945/60281292-fb818880-9904-11e9-8a6a-fc727f0baf7b.png">

There are still many missing corner cases, but it seems like I am on the right track.

I’ve also improved the DMA inspector a bit. Now it is capable of visualising Copper DMA, too.

<img width="929" alt="Screenshot 2019-06-27 at 17 39 05" src="https://user-images.githubusercontent.com/12561945/60281312-050af080-9905-11e9-858f-b4027cc0f3bb.png">

I’m going to open another repro to store the test cases there. Because there is no such thing for Amiga emulators as the VICE test suite for C64 emulators, I think it makes sense to build one up from scratch step by step.

**Comment by dirkwhoffmann** [comment]
> I feared my "story" had a negative impact on your patience after I posted it and I felt sorry.

No, not at all. As vAmiga is my second emulator (after VirtualC64), I know that writing such a thing is a marathon and the road to success is full of setbacks. My worst nightmare with VirtualC64 was to get the Emulamer's Errata Demo to work.  A few times I was about to throw my equipment against the wall and to trash the whole project. I’m still a little 😖 about the Emulamer group, because they stole so much of my life-time with their stupid cat and mouse games. And then there is Crest 🤬... Oh, I better change the subject... 

I better talk about why I am doing this at all. Actually, there are three distinct reasons. The first is simple: Writing an emulator is great fun 🥳. One feels a little bit like Frankenstein who brings something dead (a simple binary Rom loaded from the internet) to life. The second reason is that I need a little change from time to time. In the last 15 years I've spent a lot of my time writing books (which is my true passion and a lot of fun, too). What has happened is that after finishing my ninth book I suddenly run out of ideas (which is a pity, because at some point in my life, likely after too many 🍺, I decided to write at least 10 😦). Unfortunately, I had no idea what a new book could be about, so I decided to do some programming until such an idea is coming. The third reason is also simple: I want to stay in touch with latest technology. I left the software industry in 2004 and since then there have been great developments. By writing VirtualC64, I learned how to code in Swift (which wasn't there at all in 2004), how to write macOS apps, how to optimize with Intel Intrinsics and SSE, and lots of other fun stuff. Not to forget: By hacking some checksum computation code into SAE to compare my Blitter results with yours, I even learned a little JavaScript 👍. And last but not least: As there are always students who believe their teachers can't code, it's pleasurable to have the ability to prove them wrong 😎. 

Overall, I am curious myself how far the project will go. The biggest risk for vAmiga is probably that I will soon have an idea for a new book 🤔.  We'll see...
